### PR TITLE
Always generate avatar

### PR DESCRIPTION
**Pull request by rullzer**
@LukasReschke as discussed

Even if no avatar is set we should just generate the image. This to not
duplicate the code on all the clients. And only server images from the
avatar endpoint.

Todo:
- [x] Save avatar
- [x] Move this to the actual avatar code instead of the Controller

Future work:
- use imgsrc so the browser fetches the right size on hdpi stuff
- extend caching

### CODE REVIEW

**Comment by rullzer** [comment]
Ah mmm so upating the JS is a tiny bit tricky as there is also some code to handle unknown users etc.

**Comment by rullzer** [comment]
So @tobiasKaminsky this is what you wanted right?

@jancborchardt please checkout this branch and try to share etc. To see if there are no weird avatar hickups now.

**Comment by tobiasKaminsky** [comment]
@rullzer this is great. So we can drop avatar generation once NC 12 has end of life :-)

File: core/Controller/AvatarController.php [hunk 1]
```diff
@@ -112,10 +112,12 @@ public function __construct($appName,
 		$this->timeFactory = $timeFactory;
 	}
 
+
+
+
 	/**
 	 * @NoAdminRequired
 	 * @NoCSRFRequired
-	 * @NoSameSiteCookieRequired
```
Comment (LukasReschke) [inline core/Controller/AvatarController.php]: Why did you remove this? This is for example required when Collabora is embedding avatars since it runs under a different origin :)

File: lib/private/Avatar.php [hunk 2]
```diff
@@ -210,4 +230,160 @@ private function getExtension() {
...
+		}
+
+		// Reduce values bigger than rgb requirements
+		$rgb[0] %= 255;
+		$rgb[1] %= 255;
+		$rgb[2] %= 255;
+
+		$hsl = $this->rgbToHsl($rgb[0], $rgb[1], $rgb[2]);
+
+		// Classic formulla to check the brigtness for our eye
```
Comment (LukasReschke) [inline lib/private/Avatar.php]: /formulla/formula/

File: lib/private/Avatar.php [hunk 3]
```diff
@@ -210,4 +230,160 @@ private function getExtension() {
...
+		}
+
+		return $this->hslToRgb($hsl[0], $sat, $lum);
+	}
+
+	/**
+	 * @param double $h Heu in range [0, 1]
+	 * @param double $s Saturation in range [0, 1]
+	 * @param double $l Lightness in range [0, 1]
+	 * @return int[] Array containging r g b in the range [0, 255]
```
Comment (LukasReschke) [inline lib/private/Avatar.php]: /containging/containing/

File: lib/private/Avatar.php [hunk 4]
```diff
@@ -210,4 +230,160 @@ private function getExtension() {
...
+		$bright = sqrt(0.299 * ($rgb[0] ** 2) + 0.587 * ($rgb[1] ** 2) + 0.114 * ($rgb[2] ** 2));
+		if ($bright >= 200) {
+			$sat = 0.60;
+		}
+
+		return $this->hslToRgb($hsl[0], $sat, $lum);
+	}
+
+	/**
+	 * @param double $h Heu in range [0, 1]
```
Comment (LukasReschke) [inline lib/private/Avatar.php]: /Heu/Hue/ maybe?

File: lib/private/Avatar.php [hunk 5]
```diff
@@ -210,4 +230,160 @@ private function getExtension() {
 		}
 		throw new NotFoundException;
 	}
+
+	private function generateAvatar($userDisplayName, $size) {
```
Comment (LukasReschke) [inline lib/private/Avatar.php]: PHPDoc on the parameters would be awesome 😄

**Review by LukasReschke — DISMISSED** [review]

File: core/Controller/AvatarController.php (hunk 1 above)
Comment (rullzer) [reply core/Controller/AvatarController.php]: Good question

**Review by MorrisJobke — APPROVED** [review]
I made the font use semibold weight and made it a bit better centered, because it was slightly off.

**Comment by MorrisJobke** [comment]
bildschirmfoto 2017-12-08 um 16 24 04

The only thing that may be a bit different is, that we now always have an avatar in the top right - cc @nextcloud/designers

**Comment by skjnldsv** [comment]
I'm OK with that. This is also pretty nice for the top right menu list drop-down positioning issue we have here and there :)

**Comment by MariusBluem** [comment]
I like always having the avatar at the top 👍

**Comment by MorrisJobke** [comment]
Discussed with @LukasReschke on IRC and good to get in.

**Comment by MorrisJobke** [comment]
@rullzer Seems to break the integration tests for sharing :/ See @danxuliu's comment in https://github.com/nextcloud/server/pull/6715#issuecomment-350548820

**Comment by rullzer** [comment]
I know. I have a patch for that already

**Comment by MorrisJobke** [comment]
> I know. I have a patch for that already

#7342

**Comment by jancborchardt** [comment]
Hm, in the past @karlitschek and I discussed this with avatar placeholder in top right and we didn’t really like it as it’s very present and colorful. However it kind of nudges you to set a proper image, so I guess it’s fine after all. :)

**Comment by jancborchardt** [comment]
Actually this causes a confusion with new users or installations. Cause in the top right the avatar is shown and not the settings icon, it is not immediately obvious where the settings are:

https://github.com/nextcloud/server/issues/7535

(When you are used to the UI and added a picture, you also know where the settings are.)

**Comment by MariusBluem** [comment]
I disagree with you @jancborchardt ... this behaviour is known by the users from so many services ... not only by Nextcloud.

To make it crystal clear we could show some arrows with descriptions in the firstrunwizard :)

**Comment by jancborchardt** [comment]
> To make it crystal clear we could show some arrows with descriptions in the firstrunwizard :)

If it's known, we don't need a pointer for such an obvious menu in the first run wizard. ;)

**Comment by MorrisJobke** [comment]
> Actually this causes a confusion with new users or installations. Cause in the top right the avatar is shown and not the settings icon, it is not immediately obvious where the settings are:

I guess the main problem is, that we don't have an indicator for menus. On github they put for example a little arrow beside the avatar to indicate the menu: 

bildschirmfoto 2017-12-16 um 12 32 18

**Comment by MariusBluem** [comment]
**Dropbox:**
c3d27750-4e64-44d5-a824-f1ec369ab2b5

**Microsoft:**
dfcc7cd6-ef3d-4890-8162-c26912e8ee1f

No indicator ;)
